### ISSUE TITLE
.nrepl-port file  ===> ClojureShadowSelect errors

### ISSUE THREAD

**Issue by thecontinium** [Olical/conjure#584]
Hi 

I have an  nrepl running which has a .nrepl-port.
When I delete the session it automatically start a new session on the same port.

```
; Sponsored by @valsen ❤.    <====== open clojure file and show log ( only shows the )
; -----------------------------------------------------------  <======= Close all repl sessions here
; Closed all sessions (1)
; -----------------------------------------------------------
; Assumed session: American Eskimo Dog (Clojure).           <<<<<<<<<<<<<<< New session automatically started
; ----------------------------------------------------------   <====== list the sessions
; Sessions (1):
; >1 - American Eskimo Dog (Clojure)
```

This is also the same behaviour with list sessions. It automatically connects when there is a port file.

```
; Sponsored by @edlandm ❤                                      <====== open clojure file and show log ( only shows this )
; -----------------------------------------------------------  <====== list the sessions
; localhost:56328 (connected): .nrepl-port                 <<<<<<<<<<<<<<< New Session automatically started
; -----------------------------------------------------------
; Assumed session: Sealyham Terrier (Clojure)
; ----------------------------------------------------------- 
; Sessions (1):
; >1 - Sealyham Terrier (Clojure).                                              

```

I want to be able to create a new client state, delete all sessions and start up a shadow clojure repl.

This auto starting when port files exists ( I think ) is having side effects with starting up shadow-clj with `ClojureShadowSelect ui` where I am getting this error only when I have a .nrepl-port file.

```
; --------------------------------------------------------------------------------
; shadow-cljs (select): app
; (err) Execution error (ClassNotFoundException) at java.net.URLClassLoader/findClass (URLClassLoader.java:471).
; (err) shadow.cljs.devtools.api

```

 If I start up the `shadow-clj watch ui` and `ClojureShadowSelect ui` on it's own without an .nrepl-port file it is fine

I have:
```
vim.g['conjure#client_on_load'] = false
vim.g['conjure#client#clojure#nrepl#connection#auto_repl#enabled'] = false
```

**Comment by Olical** [comment]
This is tricky because I can't perform evaluations without a session... I think? So I always need one ready to go inside the nREPL and it's ID stored in memory.

I think if I let you get into a state with no session bad things can happen if you try to eval. I know specifically in your case you know you won't, but that's not the case usually.

Maybe the best approach (if you're scripting things) is to help you delete all nREPL sessions without having Conjure create a new one for you. I don't think that should be the default behaviour though?

Are you absolutely sure you need to have all the nREPL's sessions deleted? I don't see why that would be a requirement to start up shadow-cljs right now.

And just to check, there isn't some confusion here around nREPL sessions and connections is there? Because you have ONE nREPL server / connection / port that contains MANY sessions which are really like separate evaluation threads you can use inside the same REPL.

**Comment by thecontinium** [comment]
Hi - I think you are right about getting into a state of no sessions. These settings seem to stop any evaluations.

``` lua
vim.g['conjure#client_on_load'] = false
vim.g['conjure#client#clojure#nrepl#connection#auto_repl#enabled'] = false
vim.g['conjure#client#clojure#nrepl#connection#port_files' = '[]'

```
I didn't quite realise the interaction between the repl, nrepl+sessions, clojure script repl  and piggiebacking etc.
I had a [steer](https://clojurians.slack.com/archives/CK143P6D7/p1719017456237059?thread_ts=1718983899.085539&cid=CK143P6D7) on slack that has helped my understanding.   
I think I was trying to `:ConjureShadowSelect app` when not connected to the shadow-cljs nrepl. 


The following works for me allowing the clj and cljs repl's to use their own ( independent ) deps aliases:

-  start a nrepl from the command line using my clj deps aliases: `clojure "-M:conjure:dev:clj:clj-dev"`
-  start a shadow clojure nrepl from the command line: `shadow-cljs watch app`  with the deps aliases defined in `.shadow-cljs.edn` `{:deps {:aliases [:dev :cljs :cljs-dev :cljs-build]}`
-  save the `clj` state `:ConjureConnect <clojure-port>` `:ConjureClientState clj`
-  save the `cljs` state `:ConjureConnect <shadow-clj-port>` `:ConjureShadowSelect app`  `:ConjureClientState cljs`
-  use `:ConjureClientState` to move between repl's

I am using two different nrepl's rather than using sessions because I want them to have different class paths and not "interfere" with each other.

I'm not sure there is a way to do this with nrepl sessions and piggiebacking ?

Thanks

**Comment by Olical** [comment]
That final set of steps looks correct to me! I think you end up with a
Clojure nREPL for the server then another for shadow-cljs. Then the shadow
connect command enables the proxy through to CLJS land.

The Clojure client state usage looks correct to me, it was designed for
this problem specifically exactly how you're using it. Then you can either
set up mappings or autocommands that flip your client state key depending
on the directory you're current buffer resides in for automatic REPL
selection.

**Comment by thecontinium** [comment]
Thanks for the confirmation @Olical 

I think my initial confusion came from thinking that 
`ConjureShadowSelect app` actually did both of these steps for you

1.  `:ConjureConnect <shadow-clj-port> `
2. `:ConjureShadowSelect app  :cljs`

Not sure it warrants a new command `:ConjureConnectShadow app` but you might like to add example usage in the docs ?

I’ll close this. Thanks again.

**Comment by Olical** [comment]
Did you see this wiki page? Maybe it's not as helpful as I hoped https://github.com/Olical/conjure/wiki/Quick-start:-ClojureScript-(shadow-cljs)

**Comment by thecontinium** [comment]
I did and it helped me some of the way but I came unstuck when I wanted to run both a clojure and clojure script repl.
It does require a much better understanding of the relationship between repl(clj and cljs), nrepl+sessions, shadow-clj, figwheel and the idea of piggiebacking. [This](https://blog.agical.se/en/posts/shadow-cljs-clojure-cljurescript-calva-nrepl-basics/) helped a lot.

For those new to ClojureScript and Conjure I think an addition section to your [wiki](https://github.com/Olical/conjure/wiki/Quick-start:-ClojureScript-(shadow-cljs)) detailing running a clj repl along side a cljs repl would be useful. Something along the lines like I detailed above showing the commands to set them up and easily move between them.
